src/pybind/mgr/cephadm/schedule.py

```diff
@@ -135,67 +135,52 @@ def place(self):
-        for daemon in self.daemons:
-            if daemon.is_active:
-                for h in hosts:
-                    if h.hostname == daemon.hostname:
-                        active_hosts.append(h)
-        # remove duplicates before returning
-        return list(dict.fromkeys(active_hosts))
-
-    def prefer_hosts_with_active_daemons(self, hosts: List[HostPlacementSpec], count: int) -> List[HostPlacementSpec]:
-        # try to prefer host with active daemon if possible
-        active_hosts = self.get_hosts_with_active_daemon(hosts)
-        if len(active_hosts) != 0 and count > 0:
-            for host in active_hosts:
-                hosts.remove(host)
-            if len(active_hosts) >= count:
-                return self.scheduler.place(active_hosts, count)
-            else:
-                return list(merge_hostspecs(self.scheduler.place(active_hosts, count),
-                                            self.scheduler.place(hosts, count - len(active_hosts))))
-        # ask the scheduler to return a set of hosts with a up to the value of <count>
-        return self.scheduler.place(hosts, count)
+        # sort candidates into existing/used slots that already have a
+        # daemon, and others (the rest)
+        existing_active: List[HostPlacementSpec] = []
+        existing_standby: List[HostPlacementSpec] = []
+        others = candidates
+        for d in self.daemons:
+            hs = d.get_host_placement()
+            for i in others:
+                if str(i) == str(hs):
+                    others.remove(i)
+                    if d.is_active:
+                        existing_active.append(hs)
+                    else:
+                        existing_standby.append(hs)
+                    break
+        logger.debug('Hosts with existing active daemons: {}'.format(existing_active))
+        logger.debug('Hosts with existing standby daemons: {}'.format(existing_standby))
+
+        # Put active daemons at the front of the list
+        existing = existing_active + existing_standby
+
+        # The number of new slots that need to be selected in order to fulfill count
+        need = count - len(existing)
+
+        # we don't need any additional placements
+        if need <= 0:
+            del existing[count:]
+            return existing
+
+        # exclusive to daemons from 'mon' and 'ha-rgw' services.
+        # Filter out hosts that don't have a public network assigned
+        # or don't allow virtual ips respectively
+        if self.filter_new_host:
+            old = others
+            others = [h for h in others if self.filter_new_host(h.hostname)]
+            for h in list(set(old) - set(others)):
+                if self.spec.service_type == 'ha-rgw':
+                    logger.info(
+                        f"Filtered out host {h.hostname} for ha-rgw. Could not verify host allowed virtual ips")
+            logger.debug('filtered %s down to %s' % (old, others))
```

Comment:
this is unfortunately completely untested. But we can't do anything about it right now.